### PR TITLE
Feature/statistics

### LINKED ISSUE #16: Add statistics page
Add page with statistics which shows stats group by game type (single, PvE, PvP) and/or field size: player name (required), score (required), time of the game (?), count of turns (?)

### LINKED ISSUE #18: Add statistics service
Add a service for gathering information about finished games

### PR DESCRIPTION
**Pull request by llmaxxxll**
Closed #16 
Closed #18

### CODE REVIEW

File: dice-game-ui-angular/src/app/statistics/statistics.component.html [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<h1 align="center" class="text-muted">Statistics</h1>
+<div *ngIf="statistics.length>0; else noStatistics">
+    <table mat-table [dataSource]="statistics" class="mat-elevation-z8 width-100">
+
+        <!-- Player Column -->
+        <ng-container matColumnDef="playerId">
```
[the remark this answers is not in the document]
Comment (llmaxxxll) [reply dice-game-ui-angular/src/app/statistics/statistics.component.html]: As I understand, it's a great idea but it works only for simple columns. if you want to at least a pretty column name, it doesn't work (

File: dice-game-ui-angular/src/app/services/statistics-service.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import { GameEngine } from 'engine';
+
+export class StatisticsData {
+  gameGuid: string;
+  playerId: string;
+  totalScore: number;
+  gameType: string;
+  finishedDate: Date;
+
+  constructor(gameGuid: string, playerId: string, totalScore: number, gameType: string) {
```
[the remark this answers is not in the document]
Comment (llmaxxxll) [reply dice-game-ui-angular/src/app/services/statistics-service.ts]: Fixed

File: dice-game-ui-angular/src/app/services/local-statistics-service.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import { Guid } from 'guid-typescript';
+import { StatisticsService, StatisticsData } from './statistics-service';
+import { GameEngine } from 'engine';
```
[the remark this answers is not in the document]
Comment (llmaxxxll) [reply dice-game-ui-angular/src/app/services/local-statistics-service.ts]: Reordered

File: dice-game-ui-angular/src/app/statistics/statistics.component.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (sponkrashin) [reply dice-game-ui-angular/src/app/statistics/statistics.component.html]: Ok, keep it as is.

File: dice-game-ui-angular/src/app/statistics/statistics.component.ts [hunk 4]
```diff
@@ -1,27 +1,50 @@
 import { Component, OnInit } from '@angular/core';
 
-import { StatisticsData, StatisticsService } from '../services/statistics-service';
+import { GameStatistics, StatisticsService } from '../services/statistics-service';
+
+export class StatisticsDTO {
```
Comment (sponkrashin) [inline dice-game-ui-angular/src/app/statistics/statistics.component.ts:5]: DTOs are not needed to be classes, interface is enough to just store some data.

File: dice-game-ui-angular/src/app/statistics/statistics.component.ts [hunk 5]
```diff
@@ -1,27 +1,50 @@
...
-      .sort((game1, game2) => (game1.finishedDate < game2.finishedDate ? 1 : -1));
+      .getPlayerStatistics(curPlayer)
+      .sort((game1, game2) => (game1.creatingDate < game2.creatingDate ? 1 : -1))
+      .map(
+        (s) =>
+          new StatisticsDTO(
+            s.gameType,
+            `${s.fieldSize.width} x ${s.fieldSize.height}`,
+            s.playersStaistics?.find((ps) => ps.playerId === curPlayer)?.turnsCount ?? 0,
+            s.playersStaistics?.find((ps) => ps.playerId === curPlayer)?.score ?? 0,
```
Comment (sponkrashin) [inline dice-game-ui-angular/src/app/statistics/statistics.component.ts:44]: You do this twice `s.playersStaistics?.find((ps) => ps.playerId === curPlayer)`. Introduce a variable for this.

File: dice-game-ui-angular/src/app/statistics/statistics.component.ts [hunk 6]
```diff
@@ -1,27 +1,50 @@
...
+            s.winnerPlayer === curPlayer,
+            s.winnerPlayer !== undefined && s.winnerPlayer !== ''
```
Comment (sponkrashin) [inline dice-game-ui-angular/src/app/statistics/statistics.component.ts:46]: `!!s.winnerPlayer` is enough.

File: dice-game-ui-angular/src/app/game/game.component.ts [hunk 7]
```diff
@@ -57,35 +54,32 @@ export class GameComponent implements OnInit, AfterViewInit {
...
+      return;
     }
 
     this.gameEngine.registerOnStateChanged((engine) => {
       this.gameStorageService.saveGame(this.gameEngine, this.gameGuid);
+      // saving statistics
+      for (const player of engine.players) {
+        this.statisticsService.saveTurn(
+          this.gameGuid,
+          engine.players[0].playerId ?? 'local player',
```
[the remark this answers is not in the document]
Comment (llmaxxxll) [reply dice-game-ui-angular/src/app/game/game.component.ts]: Fixed

File: dice-game-ui-angular/src/app/statistics/statistics.component.ts (hunk 4 above)
Comment (llmaxxxll) [reply dice-game-ui-angular/src/app/statistics/statistics.component.ts]: Changed

File: dice-game-ui-angular/src/app/statistics/statistics.component.ts (hunk 5 above)
Comment (llmaxxxll) [reply dice-game-ui-angular/src/app/statistics/statistics.component.ts]: Changed

File: dice-game-ui-angular/src/app/statistics/statistics.component.ts (hunk 6 above)
Comment (llmaxxxll) [reply dice-game-ui-angular/src/app/statistics/statistics.component.ts]: Fixed

**Pull request merged**